artifact/v3/version/version.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	int version {-1};
+	string format {"unknown"};
+
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
```

Comment:
We copy the data into a buffer only to create a string from it here and throw the buffer away. It would be needed if `StringFromByteVector` took a reference and potentially modified the given buffer, but it is (I believe) a `const` reference and so we can safely pass in a buffer without creating a copy of it.

Reply:
The absolute best would be if `json` took a `Reader`, and I wouldn't have to do any of this.

Since ur intimately familiar with this. Would it be a lot of work to add you think?

Reply:
Yeah, I think I can do that. The JSON stuff we have predates our IO stuff so it wasn't possible back then. The only complication is that to do it right, we should have a UTF8Reader because bytes shouldn't be passed into the JSON parser that expects characters.

Reply:
Great :tada: :tada: